docs/development/best-practices.md

```diff
@@ -0,0 +1,125 @@
+
+### Iterating objects & containers
+
+Use `for ( ... of ...)` loops instead of `[Array|Set|Map].prototype.forEach` and `for ( ... in ...)`.
+
+- Using `for ( ... in ...)` for objects is error-prone. It will include enumerable properties from the prototype chain
+- Using `for ( ... in ...)` to iterate over arrays, will counterintuitively return the array's indices
+- Avoid `[Array|Set|Map].prototype.forEach`. It makes code harder to debug and defeats some useful compiler checks like reachability
+
+Only use `Array.prototype.map()` when the return value is used, otherwise use `for ( ... of ...)`.
+
+[Source](https://google.github.io/styleguide/tsguide.html#iterating-objects)
+
+### Exports
+
+Use named exports in all code.
+Avoid default `exports`.
+This way all `imports` follow the same pattern.
+
+[Source, reasoning and examples.](https://google.github.io/styleguide/tsguide.html#exports)
+
+### Imports
+
+Use [ES6 module](https://exploringjs.com/es6/ch_modules.html#sec_basics-of-es6-modules) syntax, i.e.
+
+```js
+import { square, diag } from 'lib';
+```
+
+or
+
+```js
+import * as lib from 'lib';
+```
+
+Avoid `require` `(as in import x = require('...');)` for imports.
+
+### HTTP & RESTful API request handling
+
+Prefer using [`got`](https://www.npmjs.com/package/got) to simplify HTTP request handling.
+
+### Dates and times
+
+Use [`Luxon`](https://www.npmjs.com/package/luxon) to handle dates and times.
+Use `UTC` to be time zone independent.
+
+[Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/distro.ts#L133-L134)
+
+### Unit testing
+
+- Use `it.each` rather than `test.each`
+- Mock Date/Time when testing a Date/Time dependent module
+  - For `Luxon` mocking see [Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/distro.spec.ts#L7-L10)
+- Prefer `jest.spyOn` for mocking single functions, or mock entire modules
+  - Avoid overwriting functions, for example: (`func = jest.fn();`)
+
+### Fixtures
+
+- Use `Fixture` class for loading fixtures
+
+```js
+Fixture.get('./file.json'); // for loading raw data
+Fixture.getJson('./file.json'); // for loading and parsing objects
+```
+
+### Working with vanilla JS files (renovate/tools only)
+
+Use [JSDoc](https://jsdoc.app/index.html) to declare types and function prototypes.
+
+[Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/tools/distro-json-generate.mjs#L7-L17)
+
+### Classes
+
+Use [Typescript getter setters (Accessors) when needed](https://google.github.io/styleguide/tsguide.html#properties-used-outside-of-class-lexical-scope).
+The getter must be a `pure function`.
+
+- Omit constructors when defining Static classes
+- [No `#private` fields](https://google.github.io/styleguide/tsguide.html#private-fields). instead, use TypeScript's visibility annotations
```

Comment:
I'm not sure the reasoning listed in your link applies to us:

> Private identifiers cause substantial emit size and performance regressions when down-leveled by TypeScript

I don't think we're down-leveling too far for this to be a problem. `#private` fields also have the benefit of being private at runtime as well.

Reply:
they have a big performance regression until node v18